contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,245 @@
+ *    private final KVStore _store;
+ *
+ *    public BatchingKVStoreClient(KVStore store) {
+ *      _store = store;
+ *    }
+ *
+ *    {@code @Override}
+ *    public void executeBatch(Integer group, Batch{@code <Long, String>} batch) {
+ *      Map{@code <Long, String>} batchResult = _store.batchGet(batch.keys());
+ *      batch.foreach((key, promise) -> promise.done(batchResult.get(key)));
+ *    }
+ *
+ *    {@code @Override}
+ *    public void executeSingleton(Integer group, Long key, BatchEntry{@code <String>} entry) {
+ *      entry.getPromise().done(_store.get(key));
+ *    }
+ *
+ *    {@code @Override}
+ *    public Integer classify(Long entry) {
+ *      return 0;
+ *    }
+ *  }
+ * </pre></blockquote>
+ *
+ * In above example there is an assumption that all keys can be grouped together. This is why method {@code classify()}
+ * trivially returns a constant {@code 0}. In practice {@code classify()} returns an equivalence class. All keys that
+ * returns equal equivalence class will constitute a batch.
+ * <p>
+ * The interaction between ParSeq and BatchingStrategy is the following:
+ * <ol>
+ *   <li>{@code batchable(String desc, K key)} is invoked to create Task instance</li>
+ *   <li>Plan is started by {@code Engine.run()}</li>
+ *   <li>When Task returned by {@code batchable(String desc, K key)} is started, the key {@code K} is remembered by a BatchingStrategy</li>
+ *   <li>When Plan can't make immediate progress BatchingStrategy will be invoked to run batchable operations:</li>
+ *   <ol>
+ *     <li>Every {@code K key} is classified using {@code classify(K key)} method</li>
+ *     <li>Keys, together with adequate Promises, are batched together based on {@code G group} returned by previous step</li>
+ *     <li>If batch contains single element, method {@code executeSingleton(G group, K key, BatchEntry<T> entry)} is invoked for it</li>
+ *     <li>If batch contains at least two elements, method {@code executeBatch(G group, Batch<K, T> batch)} is invoked for it</li>
+ *   </ol>
+ *   Both {@code executeSingleton(G group, K key, BatchEntry<T> entry)} and {@code executeBatch(G group, Batch<K, T> batch)} are invoked
+ *   in the context of their own Task instance with description given by {@code getBatchName(G group, Batch<K, T> batch)}.
+ * </ol>
+ *
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ *
+ * @param <G> Type of a Group
+ * @param <K> Type of a Key
+ * @param <T> Type of a Value
+ */
+public abstract class BatchingStrategy<G, K, T> {
+
+  private final ConcurrentHashMap<Long, BatchBuilder<K, T>> _batches =
+      new ConcurrentHashMap<>();
+
+  /**
+   * This method returns Task that returns value for a single key in allowing this strategy to batch operations.
+   * @param desc description of the task
+   * @param key key
+   * @return Task that returns value for a single key in allowing this strategy to batch operations
+   */
+  public Task<T> batchable(final String desc, final K key) {
```

Comment:
I can add this version of batchable(). It would be consistent with approach we used in Task.* API where we provide two versions of methods: one where you provide description and one where we generate it for you.